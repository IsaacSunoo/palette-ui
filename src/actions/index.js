export const setError = error => ({
  type: "SET_ERROR",
  error
});

export const setLoading = loading => ({
  type: "SET_LOADING",
  loading
});

export const setPalette = palette => ({
  type: "SET_PALETTE",
  palette
});

export const setLock = colorID => ({
  type: "SET_LOCK",
  colorID
});

export const setModal = status => ({
  type: "SET_MODAL",
  status
});

export const setProjects = projects => ({
  type: "SET_PROJECTS",
  projects
});
