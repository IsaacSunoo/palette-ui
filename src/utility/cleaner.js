export const cleanProjectsPalettes = (projects, palettes) => {
  const cleanedData = projects.map(project => {
    const { id, name } = project;
    let projectData =  { id, name, palettes: [] }
    palettes.forEach(palette => {
      if(palette.project_id === project.id){
        projectData.palettes.push(palette)
      }
    });
    return projectData;
  });
  return cleanedData;
}
