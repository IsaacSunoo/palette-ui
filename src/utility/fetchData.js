export const fetchData = async (url, body) => {
  try {
    const response = await fetch(url, body)
    const data = await response.json()
    if(!response.ok) { throw new Error(`'Fetch Call Cannot Be Made' ${data}`)}
    return data;
  } catch (error) {
    return error;
  }
}
