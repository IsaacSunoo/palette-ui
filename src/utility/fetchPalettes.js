export const fetchPalettes = async (url) => {
  try {
    const response = await fetch(url)
    if(!response.ok) { throw new Error('Fetch Call Cannot Be Made')}
    const data = await response.json()
    return data;
  } catch (error) {
    console.log(error);
    // this.props.setError('Fetch Call Cannont')
  }
}
