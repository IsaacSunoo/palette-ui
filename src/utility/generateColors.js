export const generateColors = (colorPalette) => {
  console.log(colorPalette);
  let newPalette = [];
  for(let i = 0; i < 5; i++) {
    const { color, locked } = colorPalette[i]
    if(colorPalette[i].locked) {
      newPalette.push({ color: color, locked: locked })
    } else {
      let randomColor = "#000000".replace( /0/g, function() { return (~~(Math.random()*16)).toString(16)});
      newPalette.push({ color: randomColor, locked: false })
    }
  }
  return newPalette;
}
