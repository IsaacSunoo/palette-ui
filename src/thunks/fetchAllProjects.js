import * as actions from '../actions/index';
import { fetchData } from '../utility/fetchData';
import { cleanProjectsPalettes } from '../utility/cleaner';

export const fetchAllProjects = () => {
  return async (dispatch) => {
    try {
      dispatch(actions.setLoading(true));
      const allProjects = await fetchData(process.env.REACT_APP_BACKEND_URL + '/api/v1/projects');
      const allPalettes = await fetchData(process.env.REACT_APP_BACKEND_URL + '/api/v1/palettes');
      dispatch(actions.setLoading(false));
      if(!allProjects.ok){
        throw Error(allProjects.statusText)
      }
      const combinedData = cleanProjectsPalettes(allProjects, allPalettes)
      dispatch(actions.setProjects(combinedData));
    } catch (error) {
      dispatch(actions.setError(error.message));
    }
  }
}
