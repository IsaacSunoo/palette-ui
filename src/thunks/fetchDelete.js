import * as actions from '../actions/index';
import { fetchData } from '../utility/fetchData';
import { fetchOptions } from '../utility/fetchOptions';

export const fetchDelete = (url) => {
  return async (dispatch) => {
    try {
      dispatch(actions.setLoading(true));
      const options = await fetchOptions('DELETE', {});
      await fetchData(url, options);
      dispatch(actions.setLoading(false));
    } catch (error) {
      dispatch(actions.setError(error.message));
    }
  }
}
