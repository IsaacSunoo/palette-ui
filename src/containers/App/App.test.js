import React from 'react';
import { shallow, mount } from 'enzyme'
import { App, mapStateToProps, mapDispatchToProps } from './index';

import { fetchAllProjects } from '../../thunks/fetchAllProjects';
import { generateColors } from '../../utility/generateColors';

jest.mock('../../thunks/fetchAllProjects');
jest.mock('../../utility/generateColors');

import * as actions from '../../actions/index';

const mockGeneratedPalette = [
  { color: '#fffff', locked: false },
  { color: '#00000', locked: false },
  { color: '#231f20', locked: false },
  { color: '#e6e6e6', locked: false },
  { color: '#cccccc', locked: false }
]

generateColors.mockReturnValue(mockGeneratedPalette)


describe('App', () => {

  describe('App Component', () => {
    let wrapper;
    let mockfetchAllProjects = jest.fn()
    let mockSetPalette = jest.fn()

    const mockColorPalette = [
      { color: '', locked: '' },
      { color: '', locked: '' },
      { color: '', locked: '' },
      { color: '', locked: '' },
      { color: '', locked: '' }
    ]

    beforeEach(() => {

      wrapper = shallow(<App fetchAllProjects={mockfetchAllProjects}
                             modalDisplay={false}
                             setPalette={mockSetPalette}
                             />
                        )
    })

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should fire handleEscape when handleKeydown is invoked with the code is Escape', () => {
      expect(wrapper.state()).toEqual({
        colorPalette: [
          { color: '', locked: '' },
          { color: '', locked: '' },
          { color: '', locked: '' },
          { color: '', locked: '' },
          { color: '', locked: '' }
        ]
      });
    })

    it("should invoke the method generatePalette on componentDidMount", () => {
      const instance = wrapper.instance()
      const spy = jest.spyOn(instance, 'generatePalette');

      instance.componentDidMount()

      expect(spy).toHaveBeenCalled()
    })

    it("should invoke the thunk fetchAllProjects on componentDidMount", () => {
      wrapper.instance().componentDidMount()
      expect(mockfetchAllProjects).toBeCalled()
    })

    it("should invoke setPalette with the random colors generated from generateColors method", () => {
      wrapper.instance().generatePalette()
      expect(mockSetPalette).toBeCalledWith(mockGeneratedPalette)
    })

    it("should match snapshot when modalDisplay is true", () => {
      wrapper.setProps({ modalDisplay: true })
      expect(wrapper).toMatchSnapshot()
    })

  });

  describe('mapStateToProps', () => {
    it('should return an object with a modalDisplay value', () => {
      const mockState = {
        modalDisplay: false,
        fakeState: '',
      }
      const expected = {
        modalDisplay: false,
      }

      const mappedProps = mapStateToProps(mockState)

      expect(mappedProps).toEqual(expected)
    })
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch for setPalette', () => {
      const mockData = [
        { color: '#fffff', locked: false },
        { color: '#00000', locked: false },
        { color: '#231f20', locked: false },
        { color: '#e6e6e6', locked: false },
        { color: '#cccccc', locked: false }
      ]
      const mockDispatch = jest.fn();
      const actionToDispatch = actions.setPalette(mockData)

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.setPalette(mockData)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

    it('should call dispatch for fetchAllProjects', () => {
      //Setup
      const mockData = [
        { color: '#fffff', locked: false },
        { color: '#00000', locked: false },
        { color: '#231f20', locked: false },
        { color: '#e6e6e6', locked: false },
        { color: '#cccccc', locked: false }
      ];
      const mockDispatch = jest.fn();
      const actionToDispatch = fetchAllProjects(mockData)
      //Execution
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.fetchAllProjects(mockData);
      //Expectation
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  });

});
