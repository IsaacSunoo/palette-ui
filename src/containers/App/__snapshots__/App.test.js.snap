// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App App Component should match snapshot when modalDisplay is true 1`] = `
<div
  className="App"
>
  <ConnectFunction />
  <ConnectFunction />
  <ProjectDisplay />
  <ConnectFunction />
</div>
`;

exports[`App App Component should match the snapshot 1`] = `
<div
  className="App"
>
  <ConnectFunction />
  <ConnectFunction />
  <ProjectDisplay />
</div>
`;
